crates/sui-source-validation-service/src/lib.rs

```diff
@@ -541,3 +513,48 @@ async fn api_route(
         )
     }
 }
+
+async fn check_version_header<B>(
+    headers: HeaderMap,
+    req: hyper::Request<B>,
+    next: Next<B>,
+) -> Response {
+    let version = headers
+        .get(SUI_SOURCE_VALIDATION_VERSION_HEADER)
+        .as_ref()
+        .and_then(|h| h.to_str().ok())
+        .map(|s| s.to_string());
+
+    let mut headers = HeaderMap::new();
+    headers.insert(
+        SUI_SOURCE_VALIDATION_VERSION_HEADER,
+        SUI_SOURCE_VALIDATION_VERSION.parse().unwrap(),
+    );
+
+    match version {
+        Some(v) if v != SUI_SOURCE_VALIDATION_VERSION => {
+            let error = format!(
+                "Unsupported version '{v}' specified in header \
+		 {SUI_SOURCE_VALIDATION_VERSION_HEADER}"
+            );
+            return (
+                StatusCode::BAD_REQUEST,
+                headers,
+                Json(ErrorResponse { error }).into_response(),
+            )
+                .into_response();
+        }
+        None => {
+            info!("No version set, using {SUI_SOURCE_VALIDATION_VERSION}");
+        }
+        Some(_) => (),
+    }
+    let mut response = next.run(req).await;
+    let response_headers = response.headers_mut();
+    *response_headers = headers;
```

Comment:
For composability reasons, it would be better to do the step of inserting the used version into the output header map after the request has run rather than copy the input header map over.  There may be other stuff in there that you don't want to port over, and stuff that was added in other layers of your service that would get lost by doing that.

Reply:
Done! I simultaneously ran into using and seeing `HeaderName::from_static` panic because of capitalization in the header name. No thanks to the most cryptic rust panic message. It was obnoxious to try deal with _not_ using `HeaderName::from_static` so I changed the header to be lowercase instead. The fact that [we use lowercase for the RPC header now](https://sourcegraph.com/github.com/MystenLabs/sui/-/blob/crates/sui-graphql-rpc/src/server/version.rs?L18) makes me feel fine with this shoehorning change 😄

Reply:
Indeed the lowercasing of the RPC header is for this exact reason!